### PR TITLE
several issues hit while trying to make it easy to register APIs

### PR DESCRIPTION
**Pull request by deads2k**
I was trying to create a script that would register all API versions on a given server and ended up hitting several problems.  These are the fixes.

@sttts I suspect that I won't be able to continue down the host-network approach, since that means I won't be able to use in-cluster DNS without some finagling.  It *could* be set up (and we make it work as a for instance), but the simple enablement approach will be hosted on the infrastructure.  I'll go back to that.

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -51,7 +52,7 @@ EVICTION_PRESSURE_TRANSITION_PERIOD=${EVICTION_PRESSURE_TRANSITION_PERIOD:-"1m"}
 # container bridge docker is currently using) and we don't know the IP of the
 # DNS pod to pass in as --cluster-dns. To set this up by hand, set this flag
 # and change DNS_SERVER_IP to the appropriate IP.
-ENABLE_CLUSTER_DNS=${KUBE_ENABLE_CLUSTER_DNS:-false}
+ENABLE_CLUSTER_DNS=${ENABLE_CLUSTER_DNS:-false}
```
Comment (sttts) [inline hack/local-up-cluster.sh]: Why this change?

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -405,7 +406,7 @@ function start_apiserver {
     kube::util::create_signing_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" request-header '"client auth"'
 
     # serving cert for kube-apiserver
-    kube::util::create_serving_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" "server-ca" kube-apiserver kubernetes.default.svc "localhost" ${API_HOST_IP} ${API_HOST}
+    kube::util::create_serving_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" "server-ca" kube-apiserver kubernetes.default.svc "localhost" ${API_HOST_IP} ${API_HOST} ${FIRST_SERVICE_CLUSTER_IP}
```
Comment (sttts) [inline hack/local-up-cluster.sh]: So kube-dns must have been broken for quite some time.

File: staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+dir=$(mktemp -d "${TMPDIR:-/tmp/}$(basename 0).XXXXXXXXXXXX")
+
+
+# if we have a /api endpoint, then we need to register that
+if kubectl --kubeconfig=${FROM_KUBECONFIG} get --raw / | grep -q /api/v1; then
+	group=""
+	version="v1"
+	resourceName=${version}.${group}
+	resourceFileName=${dir}/${resourceName}.yaml
+	cp ${scriptDir}/apiservice-template.yaml ${resourceFileName}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh]:
```shell
group="" \
version="v1" \
... \
eval "echo \"$(cat ${scriptDir}/apiservice-template.yaml)\"" > ${resourceFileName}
```

Of course, you have to use `${resourceName}` etc. in the template. Any kind of shell expression will do.

Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh]: This will remove the need for the GSED block.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -69,6 +70,8 @@ type APIAggregator struct {
 
 	// proxyHandlers are the proxy handlers that are currently registered, keyed by apiservice.name
 	proxyHandlers map[string]*proxyHandler
+	// handledGroups, are the groups that already have routes
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: nit: no ,

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: I was annoyed that they didn't match and our other enables aren't KUBE prefixed.  I can revert though.

File: staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh]: I've gotten used to the sed block

**Comment by deads2k** [comment]
Got the comments and added a commit to make the on infrastructure case easier.  That's probably where we start anyway.

File: staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh]: leave it if you want. But there are always some clever bash tricks to learn ;-)

File: staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+# fourth argument is reference to kube-config file that points to the aggregator you're using
+
+FROM_KUBECONFIG=${1}
+SERVICE_NAMESPACE=${2}
+SERVICE_NAME=${3}
+AGG_KUBECONFIG=${4}
+
+
+dir=$(mktemp -d "${TMPDIR:-/tmp/}$(basename 0).XXXXXXXXXXXX")
+
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/hack/register-all-apis-from.sh]: use an exit trap to delete it again

**Comment by sttts** [comment]
Commit msg of first commit is outdated I guess

**Comment by sttts** [comment]
last commit is hard to review. It's mostly a move?

**Comment by deads2k** [comment]
> last commit is hard to review. It's mostly a move?

yes

**Comment by deads2k** [comment]
@k8s-bot bazel test this
